components/gateway/cmd/main.go

```diff
@@ -124,30 +122,40 @@ func initAuditLogsSvc(done chan bool) AuditogService {
 	return auditlog.NewSink(auditlogMsgChannel)
 }
 
-func initAuditlogClient(cfg auditlog.Config) (auditlog.HttpClient, *string) {
-	if cfg.AuthMode == auditlog.Basic {
-		var basicCfg auditlog.BasicAuthConfig
-		err := envconfig.InitWithPrefix(&basicCfg, "APP")
-		exitOnError(err, "while loading basic auth config from envs")
-
-		return auditlog.NewBasicAuthClient(basicCfg), &basicCfg.Tenant
-	} else if cfg.AuthMode == auditlog.OAuth {
-		ctx := context.Background()
-
-		var oauthCfg auditlog.OAuthConfig
-		err := envconfig.InitWithPrefix(&oauthCfg, "APP")
-		exitOnError(err, "while loading oauth config from envs")
-
-		cfg := clientcredentials.Config{
-			ClientID:     oauthCfg.ClientID,
-			ClientSecret: oauthCfg.ClientSecret,
-			TokenURL:     oauthCfg.OAuthURL,
-			AuthStyle:    oauth2.AuthStyleAutoDetect,
-		}
+func initAuditlogClient(cfg auditlog.Config) (auditlog.HttpClient, *auditlog.MessageFactory) {
+	uuidSvc := uuid.NewService()
+	timeSvc := &time.TimeService{}
 
-		return cfg.Client(ctx), nil
-	} else {
-		log.Fatal(fmt.Sprintf("Invalid Auditlog Auth mode: %s", cfg.AuthMode))
-		return nil, nil
+	switch cfg.AuthMode {
+	case auditlog.Basic:
+		{
+			var basicCfg auditlog.BasicAuthConfig
+			err := envconfig.InitWithPrefix(&basicCfg, "APP")
+			exitOnError(err, "while loading basic auth config from envs")
+
+			return auditlog.NewBasicAuthClient(basicCfg), auditlog.BasicAuthMessageFactory("proxy", basicCfg.Tenant, uuidSvc, timeSvc)
+		}
+	case auditlog.OAuth:
+		{
+			ctx := context.Background()
+
+			var oauthCfg auditlog.OAuthConfig
+			err := envconfig.InitWithPrefix(&oauthCfg, "APP")
+			exitOnError(err, "while loading oauth config from envs")
```

Comment:
IMHO `exitOnError` should be used only in the main function, other functions should just return errors